### ISSUE TITLE
MultiDomain: Navigating from domain A to domain B to domain C not working and giving cross-origin frame

### ISSUE THREAD

**Issue by prashant-kashid** [cypress-io/cypress#27102]
### Current behavior

I am trying to achieve Multi Domain scenario with cypress , but currently its failing in below scenario.

Visit https://prep.comptia.org/
Click Sign In button to Navigate to Sign in Page **new domain** (https://comptiassostageb2c01.b2clogin.com/)
Click on Google Identity provider button
![image](https://github.com/cypress-io/cypress/assets/126550779/c493e106-8777-49d5-9092-137631d63851)

Here after click on Google Identity provider button, it navigate to **new domain** https://accounts.google.com/o/oauth2
Here after navigate cypress shows error as below

SecurityError
Blocked a frame with origin "https://prep.comptia.org" from accessing a cross-origin frame.
![image](https://github.com/cypress-io/cypress/assets/126550779/3922c00b-4b3d-4d5b-9d39-fdb317925c11)

### Desired behavior

Multi domain should not give error

### Test code to reproduce

**Config file**

 chromeWebSecurity: false,
  video: false,
  viewportWidth: 1200,
  viewportHeight: 800,
  experimentalModifyObstructiveThirdPartyCode: true,
  responseTimeout: 60000,
  e2e: {
    experimentalSessionAndOrigin: true,
    baseUrl: "https://prep.comptia.org",

### Cypress Version

11.2

### Node version

16

### Operating System

Windows

**Comment by MikeMcC399** [comment]
@prashant-kashid 

I suggest to check out the https://docs.cypress.io/guides/guides/cross-origin-testing guide.
This requires Cypress 12.

**Comment by prashant-kashid** [comment]
hi @MikeMcC399  MikeMcC399@mike-plummer , thank you for your guidance and support,
I have tried failure scenario with cypress 12 and below are the observations

Cypress package version: 12.15.0
Cypress binary version: 12.15.0
Bundled Node version: 16.16.0

**With implementing Cy.Origin** with Cypress 12.15
Old error of "...accessing cross origin" not appearing but we seen there is an error of 403, here Cy.Origin gets called and help to get current url which was quite helpful.

![image](https://github.com/cypress-io/cypress/assets/126550779/c48e3835-209e-4d27-8f5f-a4bb9c654e39)


Also 

I also tried adding "   experimentalModifyObstructiveThirdPartyCode : true," in cypress e2e config
by adding this 403 error get solved Google auth page opens successfully but Cy.Origin code did not trigger and cypress keeps on waiting to load the page
![image](https://github.com/cypress-io/cypress/assets/126550779/4f32218b-1101-4a85-900e-b1376040abc8)

**Comment by mike-plummer** [comment]
@prashant-kashid In order to help with `cy.origin` I would need to be able to review your spec to make sure it's being used correctly, or you would need to supply Debug Logs so I can see what Cypress is attempting to do. A 403 response indicates that a user was authenticated but lacked permissions to perform the requested activity - have you confirmed the user credentials you're using work correctly in the environment you're attempting to test?

Cypress also supports [programmatic authentication](https://docs.cypress.io/guides/end-to-end-testing/google-authentication#What-you-ll-learn) which bypasses the login form - as you've seen, using the forms can be problematic sometimes since they're highly secured and not designed to be used by automated tools to discourage malicious activity.

**Comment by prashant-kashid** [comment]
hello @mike-plummer  Thanks you for reply

I am here using Cypress with Cucumber to Achieve BDD
My Scenario is i am navigating from one Sign In page to Google O Auth page and on Google O Auth page i am trying to enter email address.
Currently all action are performed by Guest user so concern of lack permissions, 

**//this is scenario from Login.Feature file**
Scenario:2079_Verify User Sign In with Google identity provider
    Given User is on Comptia Login Page
    When User selects Google Auth

**//Here are Step definition**
Given("User is on Comptia Login Page",()=>{
  cy.visit('https://prep.comptia.org/en/');
  cy.get(".dropdown.account-dropdown").click();
})

When("User selects Google Auth",()=>{
  cy.get('.options #GoogleExchange').click(); **//Here after this click domain change to Google OAuth**
  cy.origin('accounts.google.com', () => {
    cy.url().should('include', 'accounts.google.com')
    cy.get('input[type="email"]').click(); **//Here trying to enter email on Google OAuth page**
  })
})

**Cypress Cy log**
16
getinput[type="email"]
0
AssertionError
Timed out retrying after 4000ms: Expected to find element: input[type="email"], but never found it.
[cypress/integration/step_definitions/Login/Login.js:214:6](https://prep.comptia.org/__/#)
  212 | When("User selects Google Auth",()=>{
  213 |   cy.get('.options #GoogleExchange').click(); //Here after this click domain change to Google OAuth
> 214 |   cy.origin('accounts.google.com', () => {
      |      ^
  215 |     cy.url().should('include', 'accounts.google.com')
  216 |     cy.get('input[type="email"]').click(); //Here trying to enter email on Google OAuth page
  217 |   })

![image](https://github.com/cypress-io/cypress/assets/126550779/c98f2cfb-75e6-4fe9-8212-8358010b0606)


@mike-plummer  **For this issue can we have call so that i can explain you properly.**

Also thanks for suggesting [programmatic authentication](https://docs.cypress.io/guides/end-to-end-testing/google-authentication#What-you-ll-learn)
 Will look into this.

**Comment by prashant-kashid** [comment]
Hi Any update on this ticket whether its reproducible on your end? or any new update on this?

**Comment by lmiller1990** [comment]
I am making a minimal reproduction to debug. Here's my test.

```js
describe("User is on Comptia Login Page", () => {
  it("works", () => {
    cy.visit("https://prep.comptia.org/en/");
    cy.get("a").contains("Sign in").click({ force: true });

    cy.origin("https://comptiassostageb2c01.b2clogin.com/", () => {
      // Here after this click domain change to Google OAuth
      cy.get(".options #GoogleExchange").click();
    });

    cy.origin("accounts.google.com", () => {
      cy.url().should("include", "accounts.google.com");
      cy.get('input[type="email"]').click(); //Here trying to enter email on Google OAuth page
    });
  });
});
```

I get the 403 and iframe error. Let me see how to handle this.

![Screenshot 2023-07-10 at 8 42 37 am](https://github.com/cypress-io/cypress/assets/19196536/3a7202c9-5f7f-4579-a286-b3380172b479)

**Comment by lmiller1990** [comment]
If you just want to Google Auth, there is a recipe: https://docs.cypress.io/guides/end-to-end-testing/google-authentication#Custom-Command-for-Google-Authentication

I am able to go there via `cy.visit`, but redirecting seems to 403 every time. Here is my repo: https://github.com/lmiller1990/cypress-test-tiny/compare/master...lmiller1990:cypress-test-tiny:google-auth?expand=1#diff-3c28b1e6d66b163ff3a597929e1933801408285a1b47f9f17db3b96628bd2cb3R20-R29

```js
describe("User is on Comptia Login Page", () => {
  it("does not work", () => {
    cy.visit("https://prep.comptia.org/en/");
    cy.get("a").contains("Sign in").click({ force: true });

    cy.origin("https://comptiassostageb2c01.b2clogin.com/", () => {
      // Here after this click domain change to Google OAuth
      cy.get(".options #GoogleExchange").click();
    });

    cy.origin("accounts.google.com", () => {
      cy.url().should("include", "accounts.google.com");
      cy.get('input[type="email"]').click(); //Here trying to enter email on Google OAuth page
    });
  });

  it("does not work", () => {
    cy.visit(
      "https://comptiassostageb2c01.b2clogin.com/comptiassostageb2c01.onmicrosoft.com/b2c_1a_thinkmaxoptimizely_signuporsignin/oauth2/v2.0/authorize?client_id=03f644a5-74f7-4107-979a-1cdfc255ea44&redirect_uri=https%3A%2F%2Fprep.comptia.org%2Fsignin-oidc%2Faadb2c&response_type=code&scope=openid%20profile%20offline_access&code_challenge=0_0kX22a86oBJT3LvIhRG3tsXNPcIksfhUGOOk75kK8&code_challenge_method=S256&response_mode=form_post&nonce=638245391407162870.ODViYjNhMDUtZGI5NC00N2UzLWFhNDAtMzk1MzNmZjY4OTA2YWY2ZDg0NDktMzI4MC00ZDU2LTkyY2QtYmZjNTZjNzhhZWI5&client_info=1&x-client-brkrver=IDWeb.1.25.10.0&state=CfDJ8Ox10W4xffhIrGJMUrryY7trBBbO_TJuk3wJiuGWwTEK7LgNPTf1x6X9hTT_8HJKgU9wP3MTvlQQFSHrQcrUSu_OZGIuJxbL0tQ47xWFp8P4T6MO50RTvxiBfzyxeK5cdiUVlHbEGPPgaUqbOJcKUwNKVJvsCoNj4AIEl8ptw24kJM7UOylbxBjfyRsfCUQJS1mFIVVWbXQ-B29uuUPGdkqW48WjPL3bYT2_x4kLxY-2jH1cNXAZ-5AYeXRM-OaNlwgoQxgdd2TxaaCYgy_r66y98mh3wpZPbosQEr9kmvde4YZ6DAMBo0Y_9_LaqRmpGq-5ygtl0m_5WlSl5-mvR6VhZnrkPV7h4wWKYDYmAZ7tfsZZabZE58McJmw3WbEC1gt0Th9KTvCCm9zKtPqJSjjMBbuuZKNJCRhCS6dvyDycRQCf3M9zyKHrzoFD4kNwLDtWTSwigrhQR88--7IWPZlTuNjuYk1JC4cYI6uplR3d&ui_locales=en-AU&x-client-SKU=ID_NET6_0&x-client-ver=6.25.1.0"
    );
    cy.get(".options #GoogleExchange").click();
    cy.origin("accounts.google.com", () => {
      cy.url().should("include", "accounts.google.com");
      cy.get('input[type="email"]').click(); //Here trying to enter email on Google OAuth page
    });
  });

  it.only("works", () => {
    cy.visit(
      "https://accounts.google.com/v3/signin/identifier?dsh=S-888953619%3A1688943034074580&client_id=345078337410-1ai4hap8uev9l1dgfidftbo6gn5622j7.apps.googleusercontent.com&o2v=1&redirect_uri=https%3A%2F%2Fcomptiassostageb2c01.b2clogin.com%2Fcomptiassostageb2c01.onmicrosoft.com%2Foauth2%2Fauthresp&response_type=code&scope=email+profile&service=lso&state=StateProperties%3DeyJTSUQiOiJ4LW1zLWNwaW0tcmM6MTNjODlmMmUtMWMxZi00YTM4LWE2YjAtMGU5NGNlZGZhMjE0IiwiVElEIjoiODJiYjEwMDctNDk5Ni00MzExLThlZGMtMTBjNjczMjQ4MDVmIiwiVE9JRCI6IjJkNDlmNjdhLWUyYzUtNGRmMS05ZDI2LTE0YjQzZTAxZGVkOCJ9&flowName=GeneralOAuthFlow&continue=https%3A%2F%2Faccounts.google.com%2Fsignin%2Foauth%2Fconsent%3Fauthuser%3Dunknown%26part%3DAJi8hAOFOcvLj6hxXinQfOsM4IRa0RS7419DwqxPP8ve8s_USYjyo3Mqt8z822W9QNXtvWiTUjzA6x_ZB9lub3CiLzeELbQzAro3gwV0D8_2CdXoz8GQ4BdvKu5yyk8M04ImA9zj01kKeUso6t0h35CyWjiqv9QGR7GGJ-8K9RtZrsj-9Hg5LqRKUw-o8MQSUS4GFJrvcyYOih2oCeUSEDY6AzW7CDvZlMjx2ql_0VQUxyYOxYzpOkf0E2lvTDHywheEdDiC1pg8Uu1HpvQN5u-seu_h14MLbMtA-eYo5olmYScc0WHWJwBxhA7jWBDhFQ6CXEIm986Lvv_CMv84_7K-EIzrYGf_oChaepwdMnc69DPkVVQsVeNh76iVXM-Ea6eNXh83QunC51IAUoxKh1yRO3E483KcC1XnCcM_kPxGwq6ydNLpX3_Ov_bskS-1T6Xg7p5pZEO_2kt9X1CkKgB8Gwn-2I3bO87U67yW30XeBdpe5W8mG_4%26as%3DS-888953619%253A1688943034074580%26client_id%3D345078337410-1ai4hap8uev9l1dgfidftbo6gn5622j7.apps.googleusercontent.com%23&app_domain=https%3A%2F%2Fcomptiassostageb2c01.b2clogin.com&rart=ANgoxce93HCTbfIrLOoR1BehXUmTPGU_ASEgR3LuyazsbSnCOF2SBgC5wIswQTmulIxyvbQx7ZLFq3JWM8d0SwEGzfqskAuX3A"
    );
    cy.url().should("include", "accounts.google.com");
    cy.get('input[type="email"]').click(); //Here trying to enter email on Google OAuth page
  });
});
```

Related: 
- https://github.com/cypress-io/cypress/issues/23814
- https://github.com/cypress-io/cypress/issues/22987

**Comment by lmiller1990** [comment]
Okay, I got it to work @prashant-kashid. Working example:

https://github.com/lmiller1990/cypress-test-tiny/compare/master...lmiller1990:cypress-test-tiny:google-auth-working?expand=1

I needed to add `experimentalModifyObstructiveThirdPartyCode`. More info [here](https://docs.cypress.io/guides/references/experiments).


```js
const { defineConfig } = require('cypress');

module.exports = defineConfig({
  e2e: {
    experimentalModifyObstructiveThirdPartyCode: true,
  }
});
```

I hope this helps solve you issue. Let me know if you run into any others.
